File: go/basic/responsibility/v1/responsibility.go
package v1

import "fmt"

type Person struct {
	name   string
	dayoff int
	reason string
	leader Manager
}

func (p Person) getName() string {
	return p.name
}

func (p Person) getDayOff() int {
	return p.dayoff
}

func (p Person) getReason() string {
	return p.reason
}

func (p *Person) setLeader(leader Manager) {
	p.leader = leader
}

func (p Person) reuqest() {
	fmt.Println(p.name, "申请请假", p.dayoff, "天。请假事由：", p.reason)
	if p.leader != nil {
		p.leader.handleRequest(p)
	}

}

type Manager interface {
	getName() string
	getTitle() string
	setNextHandler(Manager)
	getNextHandler() Manager
	handleRequest(Person)
}

type Supervisor struct {
	name        string
	title       string
	nexthandler Manager
}

func (s Supervisor) getName() string {
	return s.name
}

func (s Supervisor) getTitle() string {
	return s.title
}

func (s *Supervisor) setNextHandler(nexthandler Manager) {
	s.nexthandler = nexthandler
}

func (s Supervisor) getNextHandler() Manager {
	return s.nexthandler
}

func (s Supervisor) handleRequest(p Person) {
	if p.getDayOff() <= 2 {
		fmt.Println("同意", p.getName(), "请假，签字人：", s.getName(), "(", s.getTitle(), ")")
	}
	nextHandler := s.getNextHandler()
	if nextHandler != nil {
		nextHandler.handleRequest(p)
	}
}

type DepartmentManager struct {
	name        string
	title       string
	nexthandler Manager
}

func (dm DepartmentManager) getName() string {
	return dm.name
}

func (dm DepartmentManager) getTitle() string {
	return dm.title
}

func (dm *DepartmentManager) setNextHandler(nexthandler Manager) {
	dm.nexthandler = nexthandler
}

func (dm DepartmentManager) getNextHandler() Manager {
	return dm.nexthandler
}

func (dm DepartmentManager) handleRequest(p Person) {
	if p.getDayOff() > 2 && p.getDayOff() <= 5 {
		fmt.Println("同意", p.getName(), "请假，签字人：", dm.getName(), "(", dm.getTitle(), ")")
	}
	nextHandler := dm.getNextHandler()
	if nextHandler != nil {
		nextHandler.handleRequest(p)
	}
}

type CEO struct {
	name        string
	title       string
	nexthandler Manager
}

func (ceo CEO) getName() string {
	return ceo.name
}

func (ceo CEO) getTitle() string {
	return ceo.title
}

func (ceo *CEO) setNextHandler(nexthandler Manager) {
	ceo.nexthandler = nexthandler
}

func (ceo CEO) getNextHandler() Manager {
	return ceo.nexthandler
}

func (ceo CEO) handleRequest(p Person) {
	if p.getDayOff() > 2 && p.getDayOff() <= 22 {
		fmt.Println("同意", p.getName(), "请假，签字人：", ceo.getName(), "(", ceo.getTitle(), ")")
	}
	nextHandler := ceo.getNextHandler()
	if nextHandler != nil {
		nextHandler.handleRequest(p)
	}
}

type Administrator struct {
	name        string
	title       string
	nexthandler Manager
}

func (admit Administrator) getName() string {
	return admit.name
}

func (admit Administrator) getTitle() string {
	return admit.title
}

func (admit *Administrator) setNextHandler(nexthandler Manager) {
	admit.nexthandler = nexthandler
}

func (admit Administrator) getNextHandler() Manager {
	return admit.nexthandler
}

func (admit Administrator) handleRequest(p Person) {
	fmt.Println("同意", p.getName(), "请假，签字人：", admit.getName(), "(", admit.getTitle(), ")")
}

File: cpp/responsibility/Responsibility.cpp
/**
 * Author: arqqqq
 * Date:   2022/4/13
 */

#include "Responsibility.h"

LeaveTips::LeaveTips(const std::string &name, const std::string &reason, int leave_day):__name(name),
                                                                                        __reason(reason),
                                                                                        __leave_days(leave_day){
    __leader = nullptr;
}

void LeaveTips::setLeader(std::shared_ptr<Manager> leader) {
    __leader = std::move(leader);
}

std::string LeaveTips::getName() const { return __name; }

std::string LeaveTips::getReason() const { return __reason; }

int LeaveTips::getDays() const { return __leave_days; }

void LeaveTips::request() {
    printf("%s apply for leave %d day. leave reason: %s\n", __name.c_str(), __leave_days, __reason.c_str());
    if( __leader != nullptr ){
        __leader->handleRequest(*this);
    }else{
        printf("No one handle this application!\n");
    }
}

Manager::Manager(const std::string &name, const std::string &company):__name(name),
                                                                     __company(company){
    nextManager = nullptr;
}

std::string Manager::getName() const { return __name; }

std::string Manager::getCompany() const { return __company; }

void Manager::setNextHandler(std::shared_ptr<Manager> handle) {
    nextManager = std::move(handle);
}

std::shared_ptr<Manager> Manager::getNextHandler() const { return std::move(nextManager); }

Supervisor::Supervisor(const std::string &name, const std::string &company): Manager(name, company){}

void Supervisor::handleRequest(const LeaveTips &tips) const {
    if(tips.getDays() <= 2)
        printf("Agree %s's apply, Signatories: %s Title: (%s)\n", tips.getName().c_str(), __name.c_str(), __company.c_str());
    if(nextManager != nullptr)
        nextManager->handleRequest(tips);

}

DepartmentManager::DepartmentManager(const std::string &name, const std::string &company): Manager(name, company){}

void DepartmentManager::handleRequest(const LeaveTips &tips) const {
    if(tips.getDays() > 2 && tips.getDays() <= 5)
        printf("Agree %s's apply, Signatories: %s Title: (%s)\n", tips.getName().c_str(), __name.c_str(), __company.c_str());
    if(nextManager != nullptr)
            nextManager->handleRequest(tips);
}

CEO::CEO(const std::string &name, const std::string &company): Manager(name, company){}

void CEO::handleRequest(const LeaveTips &tips) const {
    if(tips.getDays() > 5 && tips.getDays() <= 22)
        printf("Agree %s's apply, Signatories: %s Title: (%s)\n", tips.getName().c_str(), __name.c_str(), __company.c_str());
    if(nextManager != nullptr)
        nextManager->handleRequest(tips);
}

Administrator::Administrator(const std::string &name, const std::string &company): Manager(name, company){}

void Administrator::handleRequest(const LeaveTips &tips) const {
    printf("%s's leave request has been reviewed and is true! Filed for processing, Signatories: %s Title: (%s)\n\n", tips.getName().c_str(), __name.c_str(), __company.c_str());
}
